### PR TITLE
SMTChecker: Use Z3 through the SMT-LIB interface

### PR DESCRIPTION
**Pull request by blishko**
Instead of using Z3 as a library, we use Z3 as an external solver, the same way as we already use Eldarica and cvc5.

After discussion with @ekpyron, we decided that emscripten build of `solc` will be an exception.
The rationale is that switching also the javascript binary of the compiler would require changes in `solc-js` and would put additional burden on users who would like to use SMTChecker.
Fortunately, Z3 exposes an API method `Z3_eval_smtlib2_string` (from C interface of Z3), which seems to behave exactly as calling the solver on the given SMT-LIB2 script.
Thus, we can reuse the same infrastructure we have for processing answers in both cases (calling external process, or calling this API method).

### CODE REVIEW

File: libsolidity/CMakeLists.txt [hunk 1]
```diff
@@ -141,6 +141,10 @@ set(sources
 	formal/SymbolicVariables.h
 	formal/VariableUsage.cpp
 	formal/VariableUsage.h
+	formal/Z3CHCSmtLib2Interface.cpp
+	formal/Z3CHCSmtLib2Interface.h
+    formal/Z3SMTLib2Interface.cpp
+    formal/Z3SMTLib2Interface.h
```
Comment (matheusaaguiar) [inline libsolidity/CMakeLists.txt]: looks like the indentation is wrong for these.

File: libsolidity/formal/ExpressionFormatter.cpp [hunk 2]
```diff
@@ -245,7 +245,14 @@ std::optional<std::string> expressionToString(smtutil::Expression const& _expr,
 	if (smt::isNumber(*_type))
 	{
 		solAssert(_expr.sort->kind == Kind::Int);
-		solAssert(_expr.arguments.empty());
+		solAssert(_expr.arguments.empty() || _expr.name == "-", "");
```
Comment (matheusaaguiar) [inline libsolidity/formal/ExpressionFormatter.cpp]: Is this related to Z3 or just an improvement overall?
Also, since the assert error message is empty, we can omit it.
```suggestion
		solAssert(_expr.arguments.empty() || _expr.name == "-");
```

File: libsolidity/formal/Z3CHCSmtLib2Interface.cpp [hunk 3]
```diff
@@ -47,34 +62,56 @@ CHCSolverInterface::QueryResult Z3CHCSmtLib2Interface::query(smtutil::Expression
...
-		setupSmtCallback(true);
-		if (!boost::starts_with(response, "unsat"))
-			return {CheckResult::SATISFIABLE, Expression(true), {}};
-		return {CheckResult::SATISFIABLE, Expression(true), graphFromZ3Answer(response)};
-	}
+#ifdef EMSCRIPTEN_BUILD
+		z3::set_param("fp.xform.slice", true);
+		z3::set_param("fp.xform.inline_linear", true);
+		z3::set_param("fp.xform.inline_eager", true);
+		std::string response = [&](){ z3::context context; return Z3_eval_smtlib2_string(context, query.c_str()); }();
```
Comment (matheusaaguiar) [inline libsolidity/formal/Z3CHCSmtLib2Interface.cpp]: Why not just this or similar instead of the lambda?
```suggestion
		std::string response = Z3_eval_smtlib2_string(z3::context{}, query.c_str());
```

**Review by matheusaaguiar — COMMENTED** [review]
I don't see anything wrong with the code in a first pass. I am building and running tests locally to confirm everything works, just for completeness sake.

**Comment by ekpyron** [comment]
Did something change in the model checker engine settings here? I just quickly checked the emscripten binary on remix and was surprised to only get CHC responses - and then also checked the static ubuntu binary, which also only seems to run CHC no matter what I pass in as ``--model-checker-engine``. We could change the behaviour of all this, but what I just saw seemed a bit odd at least :-).

**Comment by blishko** [comment]
That should not be the case. I will investigate. What exactly have you been trying? What examples?

**Comment by ekpyron** [comment]
I just checked again: my mistake was mainly that (since that's easier for remix) using the old experimental smtchecker pragma, which seems to override the CLI settings. And also forgot that for some cases if CHC solves it, we won't even query BMC anymore.
With some more care in testing all seems to indeed work as expected! Sorry for the false alarm :-).

File: libsolidity/CMakeLists.txt (hunk 1 above)
Comment (blishko) [reply libsolidity/CMakeLists.txt]: Fixed!

File: libsolidity/formal/ExpressionFormatter.cpp (hunk 2 above)
Comment (blishko) [reply libsolidity/formal/ExpressionFormatter.cpp]: This was a bug that was dormant and was exposed by the switch to using SMT-LIB interface much more.

I've removed the unnecessary second argument now.

File: libsolidity/formal/Z3CHCSmtLib2Interface.cpp (hunk 3 above)
Comment (blishko) [reply libsolidity/formal/Z3CHCSmtLib2Interface.cpp]: Thank you for the suggestion, I haven't thought of that!
This should indeed work and is much nicer!
Please, check the updated code.

File: Changelog.md [hunk 4]
```diff
@@ -17,6 +17,7 @@ Language Features:
 
 Compiler Features:
  * General: Generate JSON representations of Yul ASTs only on demand to reduce memory usage.
+ * SMTChecker: Z3 is now a runtime dependency, not a build dependency (except for emscripten build).
```
Comment (ekpyron) [inline Changelog.md]: Changelog entry needs to move to 0.8.29 - but then I hope we can merge this soon!

Comment (blishko) [reply Changelog.md]: Changelog entry has been moved.

**Review by matheusaaguiar — DISMISSED** [review]
Just needs a rebase and to squash commits.

**Comment by blishko** [comment]
I should squash everything to a single commit?

**Comment by matheusaaguiar** [comment]
> I should squash everything to a single commit?

That's usually how we do it. But if you want/feel the need to separate things, maybe group everything in 2 or 3 major commits.

**Comment by cameel** [comment]
The comments here actually look perfectly fine to me.

It's not really about the number. What we don't want is messy history where things are changed and then changed back within the same PR, making it hard to follow. If that's the case then squashing into a single commit is a simple way out. But each commit being one logical, atomic change with a good description is actually superior if we can get it.

So just make sure there are no separate commits for review corrections (they should all be squashed into the original commits they fix) and it's fine. The one commit called "Fixup" sounded like something like that, but it seems that it's actually just normal fix so it's fine.

**Comment by blishko** [comment]
I tried to keep the commits logically compact. I think I can improve some of the commit messages and squash some of the commits together, but I would prefer to keep at least some commits separate, because they address different issues.

File: libsmtutil/Exceptions.h [hunk 5]
```diff
@@ -48,4 +48,25 @@ struct SMTLogicError: virtual util::Exception {};
 		"SMT assertion failed" \
 	)
 
+
+// Error to indicate that some problem occurred during an interaction with external solver.
+// This could be a problem with calling the solver or unexpected situation during the processing of solver's response.
+struct SMTSolverInteractionError: virtual util::Exception {};
+
+#if !BOOST_PP_VARIADICS_MSVC
+#define smtSolverInteractionAssert(...) BOOST_PP_OVERLOAD(smtSolverInteractionAssert_,__VA_ARGS__)(__VA_ARGS__)
```
Comment (cameel) [inline libsmtutil/Exceptions.h]: Please don't call it an "assert". It's a validation and that should be clear from the name. `smtSolverInteractionRequire()` would be a better name.

We already have the same problem with `assertThrow()` and `solUnimplementedAssert()` (#15139) and I was planning to rename them eventually. It would be good not to introduce even more things following this misleading pattern :)

**Review by cameel — COMMENTED** [review]
I took a quick look at the PR before we merge it and there's one last thing I think should be changed.

File: libsmtutil/Exceptions.h (hunk 5 above)
Comment (cameel) [reply libsmtutil/Exceptions.h]: Oh, and also, for validations, I think the message should be mandatory. We should always give user the reason why things failed and making it optional makes it too easy to skip it (I see that a lot of these validations here actually do not have messages and they should).

**Comment by curtwaynesamson** [comment]
Thank You

File: libsmtutil/Exceptions.h (hunk 5 above)
Comment (blishko) [reply libsmtutil/Exceptions.h]: Thanks for the comments!
I've updated the code according to your suggestions.

File: libsmtutil/CHCSmtLib2Interface.cpp [hunk 6]
```diff
@@ -27,8 +27,9 @@
 #include <boost/algorithm/string/join.hpp>
 #include <boost/algorithm/string/predicate.hpp>
 
-#include <range/v3/algorithm/find_if.hpp>
 #include <range/v3/algorithm/all_of.hpp>
+#include <range/v3/algorithm/find_if.hpp>
```
Comment (matheusaaguiar) [inline libsmtutil/CHCSmtLib2Interface.cpp]:
```suggestion

```
Not needed anymore. It was introduced earlier in this [commit](https://github.com/ethereum/solidity/commit/384f005071afe0da2e209246ee74bc20ed2cdc89), but `ranges::find_if` is not used anywhere in this file currently.

File: test/cmdlineTests/model_checker_invariants_contract_reentrancy/err [hunk 7]
```diff
@@ -1,14 +1,8 @@
...
-  | 		^^^^^^^^^^^
-
 Info: CHC: 1 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
 
 Info: Contract invariant(s) for model_checker_invariants_contract_reentrancy/input.sol:test:
-(!(x >= 1) || true)
+(true || !(x >= 1))
 Reentrancy property(ies) for model_checker_invariants_contract_reentrancy/input.sol:test:
-(((!(x <= 0) || !(<errorCode> >= 1)) && (!(x <= 0) || !(x' >= 1))) || true)
+(((<errorCode> = 0) && (x!6 = x!4) && (x' = x)) || true || true || true || true)
```
Comment (matheusaaguiar) [inline test/cmdlineTests/model_checker_invariants_contract_reentrancy/err:6]: What does `x!6 = x!4` mean?

Comment (blishko) [reply test/cmdlineTests/model_checker_invariants_contract_reentrancy/err:6]: Looks like these are variables names in Z3 that we are for some reason not translating back (to expressions like `x` and `x'`).
That is something to look into, but in general the invariants from Z3 are not very informative at the moment.
It basically always contains some expression that we are not able to translate back to Solidity language, so we currently overapproximate such parts with `true`. But that makes the invariant trivial.
See also https://github.com/ethereum/solidity/issues/13937

File: libsmtutil/CHCSmtLib2Interface.cpp (hunk 6 above)
Comment (blishko) [reply libsmtutil/CHCSmtLib2Interface.cpp]: Removed!

File: test/cmdlineTests/model_checker_invariants_contract_reentrancy/err (hunk 7 above)
Comment (matheusaaguiar) [reply test/cmdlineTests/model_checker_invariants_contract_reentrancy/err:6]: Ah, ok. I was a bit confused by the `!` and was not sure if it was a printing/translation problem or something more serious.

**Review by matheusaaguiar — APPROVED** [review]
LGTM
